SpockOpaleye/src/TenantApi.hs

```diff
@@ -0,0 +1,133 @@
+{-# LANGUAGE Arrows #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module TenantApi
+  ( create_tenant
+  , read_tenants
+  , read_tenant_by_id
+  , make_tenant
+  , remove_tenant
+  , update_tenant
+  , activate_tenant
+  , deactivate_tenant
+  ) where
+
+import Control.Arrow
+import Data.Text
+import Database.PostgreSQL.Simple (Connection)
+import DataTypes
+import GHC.Int
+import Opaleye
+import OpaleyeDef
+import RoleApi
+import UserApi
+
+create_tenant :: Connection -> Tenant -> IO (Maybe Tenant)
+create_tenant conn tenant@Tenant {tenant_id = id
+                                 ,tenant_name = name
+                                 ,tenant_firstname = first_name
+                                 ,tenant_lastname = last_name
+                                 ,tenant_email = email
+                                 ,tenant_phone = phone
+                                 ,tenant_status = status
+                                 ,tenant_ownerid = owner_id
+                                 ,tenant_backofficedomain = bo_domain} = do
+  ids <-
+    runInsertManyReturning
+      conn
+      tenantTable
+      (return
+         ( Nothing
+         , pgStrictText name
+         , pgStrictText first_name
+         , pgStrictText last_name
+         , pgStrictText email
+         , pgStrictText phone
+         , constant status
+         , toNullable . constant <$> owner_id
+         , pgStrictText bo_domain))
+      (\(id, _, _, _, _, _, _, _, _) -> id)
+  return $
+    case ids of
+      [] -> Nothing
+      (x:xs) ->
+        Just $
+        tenant
+        { tenant_id = x
+        }
+
+activate_tenant :: Connection -> TenantId -> IO GHC.Int.Int64
```

Comment:
need better type signatures for these functions. What's the point of returning an `Int64`?